src/main/java/net/minecraftforge/common/capabilities/IAttachedCapabilityProvider.java

```
@@ -0,0 +1,107 @@
+     * This method allows attached capabilities to define if they are equivalent or not.
+     * <p>
+     * Specifically, this method compares this provider to another instance of this provider
+     * on another object, or null (if it is not present on the other object).
+     * <p>
+     * The default implementation of this method always returns true, meaning this provider
+     * will be ignored during comparison.
+     * <p>
+     * If your capability must be equivalent for itemstacks to merge, you MUST override this method.
+     * 
+     * @param other Another instance of this provider from another object, or null, if it was not present.
+     * @return true if these providers are equivalent, otherwise false.
+     */
+    default boolean isEquivalentTo(@Nullable IAttachedCapabilityProvider<C, O> other)
+    {
+        return true;
+    }
+
+    /**
+     * When {@link ItemStack}s are copied, it is inefficient to have all providers be serialized,
+     * re-attached, and deserialized.  Instead, this method allows you to copy your provider to the new stack.
+     * <p>
+     * Specifically, when a parent object is copied, this method will be called for all attached providers
+     * so that they can be copied over to the newly copied object.
+     * <p>
+     * The default implementation returns null, and returning null means this capability will NOT be attached
+     * to the newly copied object.  There will also be no way to attach to the copied object.
+     * <p>
+     * If you want to attach to copied itemstacks, you MUST override this method.
+     * 
+     * @param copiedParent The newly copied parent object.
+     * @return A deep copy of this {@link IAttachedCapabilityProvider}, or null, if it should not be copied.<br>
+     *         Note that copies will not fire {@link AttachCapabilitiesEvent}, so you must use this method.
+     */
+    default @Nullable IAttachedCapabilityProvider<C, O> copy(O copiedParent)
+    {
+        return null;
+    }
```

Would it make sense to move this to a subclass such as `IItemStackCapabilityProvider`, and have the ItemStack event gather instances of `IItemStackCapabilityProvider` directly. Then you could make `copy` abstract to force implementors to override if it necessary. This is imo nicer than a silent failure to copy caps if the modders forgets it.

---

Maybe, it makes the API on the method a bit of a mess, though not impossible to resolve - since it would require changes to `AttachCapabilitiesEvent.ItemStack` that essentially makes it error if you don't attach the subclass, which would technically break the type-contract of addCapability.
Not saying it's not doable, but it might be ugly.